### ISSUE TITLE
[BUG] Warn when noexcept is ignored

### ISSUE THREAD

**Issue by da-woods** [cython/cython#5661]
### Describe the bug

We should warn when `noexcept` is ignored - just to stop people thinking they're optimizing something that they aren't

### Code to reproduce the behaviour:

```cython
cdef object xx() noexcept:
    pass
```

This also applies to memoryviews

### Expected behaviour

This should generate a warning since exceptions are always checked for functions returning an object

### OS

Linux

### Cython version

current master

### Additional context

We might have to be a bit careful about fused functions - we probably don't want to generate warnings when *some* of the return types are valid for noexcept

**Comment by da-woods** [comment]
Should also obviously only apply to a manual noexcept, not an implicit noexcept

**Comment by matusvalo** [comment]
Just a thought. Why not directly generate an error? When an object is returned the `noexcept` is not correctly used.

**Comment by djhoese** [comment]
I just want to add the context of my stackoverflow question that lead to this issue:

https://stackoverflow.com/questions/76972950/cython-returned-memoryview-is-always-considered-uninitialized

I think an error would be acceptable to me for my case as long as it is well described or links to documentation describing the situation. Especially for memoryviews where I'm not sure I completely agree with compiler directives not having an effect on memoryview slicing...but I don't know all the ins and outs of the design so an error as an alternative seems good.

**Comment by da-woods** [comment]
I think if I were starting from scratch I'd do an error. I think now it has to be a warning to avoid breaking existing working code.

**Comment by scoder** [comment]
> it has to be a warning to avoid breaking existing working code.

Yes. Let's make it a visible warning so that users can adapt, and make it clear in the message that the code is wrong and will be considered an error eventually. We should also issue that warning in a future 0.29.x release. We can still turn it into an error in 3.1, when users had time to adapt.

**Comment by djhoese** [comment]
@matusvalo It isn't clear to me from the changes in the PR, but how are memory views handled now? While your PR handles return values, the situation this issue was originally created in reference to was my stackoverflow question. In that case, I was seeing GIL operations in the error path of memoryview slicing inside a function (if I remember correctly). Or maybe I was seeing it because I was returning a memoryview...I think.

If I'm reading your PR correctly then this case is not currently tested and may not be warned against.

**Comment by matusvalo** [comment]
Yes I forgot to check memoryviews. Let me reopen the issue.

**Comment by user202729** [comment]
> We should warn when noexcept is ignored - **just to stop people thinking they're optimizing something that they aren't**

I think this is a bad idea:

* `noexcept` is not just an optimization flag, it can also be a semantic declaration — if used on a library function, it declares "this function will never throw an exception".

  If a library supplies that, you may be able to transform your code to delete a try-except clause (because the library function is guarantee to not throw an exception).

* there's nothing that prevents cython from **actually** implementing the optimization later. As such, the current warning is a pessimization where it encourage users to remove such declaration at their cost, thus not getting the future speed boost.

* in C++, in rare cases `noexcept` can actually speed up code (e.g. because of things like strong exception guarantee of vector `resize()`), but most of the time it doesn't. Still, it's not completely meant to be a optimization guarantee like `__attribute__((always_inline))` no compiler gives a warning when no optimization is performed.
